README.md
# Django Qoqa 1.0

Django-qoqa is a command-line application to create Django's projects and to package them as Debian packages.

* Qoqa is a Zulu word which roughly translates to 'to collect'.

This project is mainly a [concept](https://standalone.co.za/blog/django_debian_package/) but can used to test the idea

This application is intended for simple traditional Django projects that don't require complex setups or deployments.


## How to install  django-qoqa
Since this project involves some debian tools and commands, the best way to use it is to install it as a debian package, that way it will install all necessary debian packaging tools.

### Prebuilt
You can get the ready made .deb package [here](https://github.com/ebsuku/python3-django-qoqa/releases), if you don't want to build it yourself.

### Source

if you want to build this package from source

* Clone this repo

* ```dpkg-buildpackage -us -uc```

* ```sudo dpkg -i python3-dajngo-qoqa_1.0.0_amd64.deb ```

Since this debian package is installed via dpkg and not via apt, dependencies will not be automatically installed, so the initial setup will fail, to install the remaining dependencies run this

* ``` sudo apt -f install ```


## Tutorial

First create a top level directory for you project, this directory will contain the django project ,the virtual environment and other files that will be needed to build a debian package later on.

```
django-qoqa new fruit
```

where `fruit` is the name of the django project


The above command will launch an interactive session asking a couple of questions


```
[qoqa] Configuring New Django Project
Available django versions
1 - 2.2 LTS
2 - 3.1
[qoqa] select django version number [1]: 
```

This allows you to select a django version, this wil show the currently supported django versions.

```
Available Databases
1 - sqlite3
2 - postgresql
Development Database [1]: 

Production Database[1]:
1 - sqlite3
2 - postgresql
```

The supported databases are SQLite3 and PostgreSQL. 

If you select PostgreSQL you'll be asked questions about the database such as database name, port, username etc.

The assumption is that you already know these details for this small project.

The production settings will be saved in a configuration file (production.cfg) with some production settings already there such as secret key and debug status.

Once you have entered the relevant details, the default packages will be installed:

* Django
* Whitenoise
* Gunicorn
* Django-debug-toolbar

## Project Layout
Once you have created the project the folder structure will look as follows

```
|-- fruit
	|-- env3
	|-- production.cfg
	|-- fruit
		|-- manage.py
		|-- fruit
			|-- settings.py
			|-- etc
		|-- fruit.cfg
```
The ```production.cfg``` contains the config settings for production.

The ```fruit.cfg``` is the configuration file that is used during development

This is an example as to how the fruit.cfg file will look

```
[STATUS]
PRODUCTION = False

[SECRET_KEY]
VALUE = "$2[Qt8a_|j+kqF-2kdbn8ii4qn)?n8\gSXr$jHA)E[ga2]`uX

[DEVELOPMENT_DATABASE]
DATABASE = sqlite3
NAME = development_db.sqlite3

[ALLOWED_HOSTS]
HOSTS = localhost
```

The production.cfg file looks somewhat similar to the fruit.cfg file, just that production will be set to True and the secret key value will be different.

The ```fruit/settings.py``` file has been modified to accommodate the fruit.cfg file.

The modified fruit/settings.py is created via the [django startproject template argument](https://docs.djangoproject.com/en/1.11/ref/django-admin/#cmdoption-startproject-template)

At this point django-qoqa has done its job in creating the project.
After this step django-qoqa is no longer involved with the django project until you need to package the project as a debian package.

So lets create a couple of django applications via the manage.py file (Don't forget to activate your virtual environment)

```
./manage.py startapp apple
./manage.py startapp banana
```


## Building the Django Project
So after some development, creating views, adding static files, now its time to build the project

First we'll need create a `requirements.txt` file

```
$pip freeze -l > requirements.txt
```
**This file needs to sit in the top level directory, same directory containing env3**

```
$django-qoqa prepare --version 0.1.0
```
**This command needs to be run in the top level directory, same directory containing env3**

The above command will generate the required files needed to begin the build process.

This is how the top level directory looks like after the build process

```
|-- fruit
	|-- env3
	|-- debian
	|-- production.cfg
	|-- MANIFEST.in
	|-- requirements.txt
	|-- setup.py
	|-- fruit
		|-- apple
		|-- banana
		|-- __init__.py
		|-- etc
```

### Debian directory
The debian directory contains the following files:

* changelog - contains documentation about all the changes made to the package.
* compat - sets he compatibility level of the build.
* control - contains details about the project eg: Descriptions, dependencies etc.
* fruit.service - This is a systemd unit file that will run the django project as a service.
* postinst - This file contains details about what should happen after a package has been installed.
* rules - This files contains the instructions as to how to build the package.
* postrm - This file contains the details to remove the project once done

If you want to read more about the debian directory and what type of files can be placed in here, you can read [this guide](https://www.debian.org/doc/manuals/maint-guide/)
packaging for debian is a bit involved so the basic configurations have been created to make things easier.


### Setup.py
This is how the default file will look

```
from setuptools import setup, find_packages

PACKAGE_DATA = {}

setup(
    name='fruit',
    version='0.1.0',
    packages=find_packages(),
    package_data=PACKAGE_DATA,
    # This is the data specified in the MANIFEST.in
    include_package_data=True,
)
```

The ```find_packages()``` function searches for all python packages that's why the ```__init__.py``` is there under the fruit directory.

We are treating the entire django project as a python package, allowing for the apple and banana apps to fall under the fruit package.


The ```PACKAGE_DATA``` dictionary is used to collect all the static files (*.css, *.js, *.djhtml) from the django apps.

So it will look like the following.


```
PACKAGE_DATA = {
	'fruit': [
		'apple/static/apple/css/*.css',
		'apple/static/apple/js/*.js'
		'apple/templates/apple/*.djhtml',
		'banana/templates/banana/*.djhtml',
		'banana/static/banana/css/*.css',
		'banana/static/banana/js/*.js'
		etc........
		
	]
}
```


More details about the setup.py file and configurations can be found [here](http://setuptools.readthedocs.io/en/latest/setuptools.html)

### Manifest.in
In the manifest file allows you to add any other data files to be included in your python package.
In this case it will contain a start_gunicorn script, this will be used by the projects systemd unit file to start and stop the project.


## Releasing the django project
So all the above files and directory (manifest,setup, debian) can be configured to suit your needs, the main idea behind this project is that you wont have to make any configuration changes or minimal ones.
We are now ready to create a .deb file

```
$django-qoqa build --version 1.0.0
```
The debian/changelog file will open up allowing you to make any final changes.
Underneath, build runs

```
dpkg-buildpackage -us -uc
```
The whole building process and .deb creation is created by [dh-virtualenv](https://github.com/spotify/dh-virtualenv)

From the dh-virtualenv github page
> dh-virtualenv is a tool that aims to combine Debian packaging with self-contained virtualenv based Python deployments.


## Installing the .deb package
Once the build process is complete a fruit_0.1.0.deb file will be created.

To install the package

```
$sudo dpkg -i fruit_0.1.0.deb
```

This will install the package in 
```
/opt/venvs/fruit/
```

To start the application
```
$systemctl start fruit
```
The production.cfg file needs to be renamed to the projects name so in this tutorial fruit.cfg


The application should now be running on 0.0.0.0:8000


Just to reiterate this is just a concept project. 
I wanted to see if it is possible to debianize a django project and what it would involve in making it work.


debian/changelog
python3-django-qoqa (1.0.0) UNRELEASED; urgency=medium
  * Enable SSL settings
  * start gunicorn in virtualenv
  * Autodiscovery of django static files
  * color coding output
  * Add start_gunicorn to data files
  * Remove manual resetting of colorama
  * Ability to create only virtualenv
  * Include template.zip in setup.py
  * choose django vserion in command line
  * Fix creating only virtualenv
  * check if project config file exists
  * check if service is running during install
  * run manage commands during service startup
  * Document pull requests
  * Remove asking for django version

 -- lunga <lunga@complex-laptop>  Tue, 24 Jul 2018 09:55:15 +0200

python3-qoqa (0.1.1) unstable; urgency=low

  * Initial public release
  * Remove project directory if pip installation failes

 -- Lunga Mthembu  <stumenz.complex@gmail.com>  Fri, 01 Aug 2017 17:00:00 +0200



qoqa/qoqa.py
import configparser
import random
import string
import os
import re

from colorama import Fore

from . import virtualenv
from . import database
from . import build

INIT_PROJECT_CONFIG = {}

# add libpq-dev
def django_version():
    """
    Specify a django version
    """
    dict_version = {"1": "2.2.*", "2": "3.1.*"}
    print(Fore.GREEN + "Avaliable django versions")
    print(Fore.GREEN + "1 - 2.2 LTS")
    print(Fore.GREEN + "2 - 3.1")
    while True:
        version = input(Fore.GREEN + "[qoqa] select django version number [1]: ")
        if version == "":
            return dict_version["1"]
        elif version not in ["1", "2"]:
            print(Fore.RED + "Invalid selection, options are 1 or 2")
        else:
            return dict_version[version]


def generate_key():
    """
    Generate Secret key for django
    """
    key = "".join(
        [
            random.SystemRandom().choice(
                string.ascii_letters + string.digits + string.punctuation
            )
            for _ in range(100)
        ]
    )
    return key


def configuration_file(status, db_config, project_name=None):
    """
    Create a configuration file based on the status(production, development)
    """
    config = configparser.RawConfigParser()
    print(Fore.GREEN + f"[qoqa] Creating {status} configuration file")

    config["SECRET_KEY"] = {"VALUE": "".join(generate_key())}
    config["ALLOWED_HOSTS"] = {"HOSTS": "127.0.0.1,localhost"}
    config["DATABASE"] = db_config
    config["STORAGE"] = {"LOCATION": "local"}

    if status == "development":
        config["STATUS"] = {"PRODUCTION": False}
        file_name = project_name

    elif status == "production":
        config["STATUS"] = {"PRODUCTION": True}
        file_name = "production"
    else:
        print(Fore.RED + "Unknown development status")
        exit()
    with open(f"{file_name}.cfg", "w") as status_file:
        config.write(status_file)
    print(Fore.GREEN + f"[qoqa] {status} configuration file created")


def create(project_name, template):
    """
    Launch interactive console and setup new project
    """
    if project_name in ["new", "build", "release"]:
        print(Fore.RED + "[qoqa] invalid project name: {}".format(project_name))
        exit()
    if os.path.isdir(project_name):
        print(Fore.RED + "[qoqa] project directory already exists")
        exit()
    print(Fore.GREEN + "[qoqa] Configuring New Django Project")
    version = django_version()
    os.mkdir(project_name)
    os.chdir(project_name)
    prod_config = database.create("production")
    dev_config = database.create("development")
    configuration_file("development", dev_config, project_name)

    os.chdir("..")
    configuration_file("production", prod_config)

    os.chdir(project_name)
    virtualenv.create(project_name, version, template)

    print(Fore.GREEN + f"[qoqa] Project {project_name} has been setup")


def new_build(version):
    """Create the debian package"""
    build.dpkg(version)


def prepare(version: str):
    """
    First check that all requirements have been met.
    Build.
    """
    project_name = os.path.basename(os.getcwd())

    print(Fore.GREEN + "[qoqa] Checking that all requirements are met.")
    if os.path.isfile("setup.py"):
        print(Fore.GREEN + "[qoqa] setup.py file exists")
    else:
        print(Fore.GREEN + "[qoqa] setup.py file does not exist, creating......")
        build.python_setup_file(project_name)

    if os.path.isfile("MANIFEST.in"):
        print(Fore.GREEN + "[qoqa] MANIFEST.in file exists")
    else:
        print(Fore.GREEN + "[qoqa] MANIFEST.in has not been created.")
        build.manifest(project_name)

    if os.path.isfile(os.path.join(project_name, "start_gunicorn")):
        print(Fore.GREEN + "[qoqa] start_gunicorn file exists")
    else:
        print(Fore.GREEN + "[qoqa] Creating new start_gunicorn script")
        build.gunicorn_file(project_name)

    if os.path.isfile(os.path.join(project_name, "__init__.py")):
        print(Fore.GREEN + "[qoqa] project __init__.py file exists")
    else:
        build.init_file(project_name)
        print(Fore.GREEN + "[qoqa]  __init__.py file created")

    if os.path.isfile("requirements.txt"):
        print(Fore.GREEN + "[qoqa] requirements.txt file exists")
    else:
        print(Fore.GREEN + "[qoqa] requirements.txt has not been created")
        build.requirements()

    if os.path.isdir("debian"):
        print(Fore.GREEN + "[qoqa] debian directory exists, updating version")
        build.update_changelog(version)
    else:
        print(Fore.GREEN + "[qoqa] debian directory does not exist")
        build.debian(version, project_name)
    print(Fore.GREEN + "[qoqa] Project ready to be built")

    print(Fore.GREEN + "[qoqa] finding django apps and updating static files")

    if build.django_app_data(project_name):
        print(Fore.GREEN + "[qoqa] project ready to be built")
    else:
        print(Fore.RED + "[qoqa] errors prepareing package")


def env(env_name, dj_version):
    """
    Only create virtualenv within project
    """
    env = virtualenv.SingleVenv(dj_version, with_pip=True)
    env.create(env_name)
